metadata/params/visitor.go

```diff
@@ -0,0 +1,267 @@
+// Copyright 2017 The kubecfg authors
+//
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+package params
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/google/go-jsonnet/ast"
+)
+
+func visitComponentsObj(component, snippet string) (*ast.Object, error) {
+	root, err := astRoot(component, snippet)
+	if err != nil {
+		return nil, err
+	}
+
+	var componentsObj *ast.Object
+	err = visit(root, &componentsObj)
+	if err != nil {
+		return nil, err
+	}
+	if componentsObj == nil {
+		return nil, fmt.Errorf("Could not find object node: %s", componentsID)
+	}
+
+	return componentsObj, nil
+}
+
+func visit(node ast.Node, componentsObj **ast.Object) error {
```

Comment:
Two things:

1. I think having 2 visitor functions is super fraught, because the Jsonnet AST is pretty complex, and we do anticipate the AST evolving significantly. I really think we should try to generalize the one we have instead of having 2.
1. I would greatly prefer to not have `**ast.Object` if we can avoid it. In the first case, it means you have to do a `nil` check for both the pointer and the pointer that's being pointed to.